### PR TITLE
make MAX_VOTERS and MAX_CANDIDATES in elections-phragmen configurable. Fix: #11092

### PR DESCRIPTION
**Pull request by sudipghimire533**
Closes: #11902 
 Polkadot companion: https://github.com/paritytech/polkadot/pull/5815

## Update
- Allow `max_voters` & `max_candidates` to be configurable from `Config` trait of `pallet_elections_phragmen`
- Put value `1000` (for `max_candidates`) & `10_000` (for `max_voters`) while creating runtime from `/bin/node/runtime`. Value taken as was in PR #11790

### CODE REVIEW

File: frame/elections-phragmen/src/lib.rs [hunk 1]
```diff
@@ -395,9 +392,10 @@ pub mod pallet {
 		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 
+			let max_candidate = <T as Config>::MaxCandidates::get();
```
Comment (kianenigma) [inline frame/elections-phragmen/src/lib.rs:395]:
```suggestion
```

File: frame/elections-phragmen/src/lib.rs [hunk 2]
```diff
@@ -395,9 +392,10 @@ pub mod pallet {
...
 			let actual_count = <Candidates<T>>::decode_len().unwrap_or(0) as u32;
 			ensure!(actual_count <= candidate_count, Error::<T>::InvalidWitnessData);
-			ensure!(actual_count <= MAX_CANDIDATES, Error::<T>::TooManyCandidates);
+			ensure!(actual_count <= max_candidate, Error::<T>::TooManyCandidates);
```
Comment (kianenigma) [inline frame/elections-phragmen/src/lib.rs:398]:
```suggestion
			ensure!(actual_count <= <T as Config>::MaxCandidates::get(), Error::<T>::TooManyCandidates);
```

**Comment by kianenigma** [comment]
Looks mostly good, need to make companions.

**Comment by sudipghimire533** [comment]
Polkadot companion: https://github.com/paritytech/polkadot/pull/5815

**Comment by Szegoo** [comment]
@sudipghimire533 Note that you have to update the ```benchmarking.rs``` file also.

**Comment by sudipghimire533** [comment]
still editing....

**Comment by sudipghimire533** [comment]
Hey mate, can you help me with this:
```
error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
   --> frame/elections-phragmen/src/benchmarking.rs:402:9
    |
402 |         let e in max_voter_count() .. max_voter_count() * MAXIMUM_VOTE as u32;
    |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`

error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
   --> frame/elections-phragmen/src/benchmarking.rs:437:9
    |
437 |         let e in max_voter_count::<T>() .. max_voter_count::<T>() * MAXIMUM_VOTE as u32;
    |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`

error: could not compile `pallet-elections-phragmen` due to 2 previous errors
```

`max_voter_count` is function which just return `T::MaxVoters::get()`

**Comment by kianenigma** [comment]
> @Szegoo maybe you can help me here:
> 
> ```
> error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
>    --> frame/elections-phragmen/src/benchmarking.rs:394:9
>     |
> 394 |         let e in T::MaxVoters::get() .. T::MaxVoters * MAXIMUM_VOTE as u32;
>     |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`
> 
> error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
>    --> frame/elections-phragmen/src/benchmarking.rs:429:9
>     |
> 429 |         let e in T::MaxVoters::get() .. T::MaxVoters::get() * MAXIMUM_VOTE as u32;
>     |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`
> ```

I think you need to wrap it in a parenthesis: https://github.com/paritytech/substrate/blob/54d6f2cbae12fa314f2bcaf6d57fba585162438f/frame/staking/src/benchmarking.rs#L815

**Comment by sudipghimire533** [comment]
Not working same error and neither can I assign it as variable and re-use. @kianenigma do you think you can help me with benchmarking.rs ?

**Comment by sudipghimire533** [comment]
> > Not working same error and neither can I assign it as variable and re-use. @kianenigma do you think you can help me with benchmarking.rs ?
> 
> I have run into the same issue when working on this, the suggested solution from @kianenigma should work. Could you give me access to your branch, so maybe I manage to get it working?

@Szegoo I have invited you to my fork. Also I tried adding `()` as well. Not good from my side. Let's see

**Comment by sudipghimire533** [comment]
> > > Not working same error and neither can I assign it as variable and re-use. @kianenigma do you think you can help me with benchmarking.rs ?
> > 
> > 
> > I have run into the same issue when working on this, the suggested solution from @kianenigma should work. Could you give me access to your branch, so maybe I manage to get it working?
> 
> @Szegoo I have invited you to my fork. Also I tried adding `()` as well. Not good from my side. Let's see

Ahh you guys meant to only wrap the `get`. Gotcha but looks like benchmark verification failing no?
```
Jul 26 20:05:52.022 ERROR runtime::elections-phragmen: Failed to run election. Number of voters exceeded    
````

**Comment by Szegoo** [comment]
@sudipghimire533 The errors are now fixed :)

**Comment by sudipghimire533** [comment]
> @sudipghimire533 The errors are now fixed :)

Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)

**Comment by Szegoo** [comment]
> > @sudipghimire533 The errors are now fixed :)
> 
> Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)

NP, note that you should put the polkadot companion in the PR description so that the companion CI checks pass. It should look like this: polkadot companion: https://github.com/paritytech/polkadot/pull/5815

**Comment by sudipghimire533** [comment]
> > > @sudipghimire533 The errors are now fixed :)
> > 
> > 
> > Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)
> 
> NP, note that you should put the polkadot companion in the PR description so that the companion CI checks pass. It should look like this: polkadot companion: [paritytech/polkadot#5815](https://github.com/paritytech/polkadot/pull/5815)

https://github.com/paritytech/polkadot/pull/5815

This looks ok?

**Comment by Szegoo** [comment]
> > > > @sudipghimire533 The errors are now fixed :)
> > > 
> > > 
> > > Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)
> > 
> > 
> > NP, note that you should put the polkadot companion in the PR description so that the companion CI checks pass. It should look like this: polkadot companion: [paritytech/polkadot#5815](https://github.com/paritytech/polkadot/pull/5815)
> 
> [paritytech/polkadot#5815](https://github.com/paritytech/polkadot/pull/5815)
> 
> This looks ok?

You should put it inside the PR description, but don't put it in a list as you did. Here is an example: https://github.com/paritytech/substrate/pull/11831

File: frame/elections-phragmen/src/lib.rs [hunk 3]
```diff
@@ -259,6 +248,14 @@ pub mod pallet {
 		#[pallet::constant]
 		type TermDuration: Get<Self::BlockNumber>;
 
+		/// The maximum number of candidates in a phragmen election.
```
Comment (Szegoo) [inline frame/elections-phragmen/src/lib.rs:251]: @kianenigma Should the docs for `MaxCandidates` and `MaxVoters` be more detailed than this?

File: frame/elections-phragmen/src/benchmarking.rs [hunk 4]
```diff
@@ -459,7 +459,7 @@ benchmarks! {
 	#[extra]
 	election_phragmen_v {
 		let v in 4 .. 16;
-		let fixed_c = MAX_CANDIDATES / 10;
+		let fixed_c = T::MaxCandidates::get() / 10;
```
Comment (kianenigma) [inline frame/elections-phragmen/src/benchmarking.rs:462]: You should be using smaller values for test, similar to the old code. 

I would do that by `let max_voters = if cfg!(test) { TEST_MAX_VOTERS } else { T::MaxVoters::get() } `

**Review by kianenigma — CHANGES_REQUESTED** [review]
Needs to be able to set smaller values for testing, otherwise running tests and benchmarks of this pallet would take ages. 

You can try yourself by doing `cargo test -p pallet-elections-phragmen --features runtime-benchmarks`

**Comment by Szegoo** [comment]
@sudipghimire533 Are you working on this, or may I finish this?

**Comment by sudipghimire533** [comment]
> @sudipghimire533 Are you working on this, or may I finish this?

I seems quite busy on my office work. You can take over. Thanks

**Comment by Szegoo** [comment]
> Needs to be able to set smaller values for testing, otherwise running tests and benchmarks of this pallet would take ages.
> 
> You can try yourself by doing `cargo test -p pallet-elections-phragmen --features runtime-benchmarks`

I don't think that is right. Please correct me if I am wrong, but in the code, the `TEST_MAX_VOTERS` and `TEST_MAX_CANDIDATES` constants are removed, so for the tests and benchmarks the values are used from the `Test` implementation of `Config`. I have put the values in the `Test` configuration to be the same as the deleted test constants. So the benchmarks don't take a long time to execute. The non-test values are specified in the runtime config and don't impact the benchmarks.

File: frame/elections-phragmen/src/benchmarking.rs (hunk 4 above)
Comment (kianenigma) [reply frame/elections-phragmen/src/benchmarking.rs:462]: what I wrote was bullshit :p

File: frame/elections-phragmen/src/lib.rs (hunk 3 above)
Comment (kianenigma) [reply frame/elections-phragmen/src/lib.rs:251]: Please add a section warning that the election happens onchain, and these numbers determine how big the election will be. Also note that if the maximum candidates are reached, we don't accept anymore. If max voters are reached, we just ignore voters.

**Review by kianenigma — APPROVED** [review]
yes, I was wrong! Please just update some docs and this is good to go.

**Comment by Szegoo** [comment]
@kianenigma I have updated the docs.

**Comment by kianenigma** [comment]
companion needs an update too.

**Comment by sudipghimire533** [comment]
> companion needs an update too.

Can you mention what needed to me updated

**Review by ruseinov — APPROVED** [review]
looks good

**Review by shawntabrizi — APPROVED** [review]

**Comment by ggwpez** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_elections_phragmen

**Review by ggwpez — APPROVED** [review]

**Comment by kianenigma** [comment]
bot rebase

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
https://github.com/sudipghimire533 @Szegoo can you both paste your Polkadot address for a tip here?

**Comment by Szegoo** [comment]
> https://github.com/sudipghimire533 @Szegoo can you both paste your Polkadot address for a tip here? 

I can't edit the PR description :P

**Comment by sudipghimire533** [comment]
Wohh I get tip as well? Cool. My address: 5CK5Wz9BRU5LaiiGAtnysAsjhTnr9ksh7rL4v9HYN1P6i1Hd

**Comment by sudipghimire533** [comment]
> > https://github.com/sudipghimire533 @Szegoo can you both paste your Polkadot address for a tip here?
> 
> I can't edit the PR description, should I provide it somehow else?

If it need to be in PR description you can put it here and I will put in the Description

**Comment by sudipghimire533** [comment]
Mine is: 1FNfKQFHFLp2Fin8Xqz1KhtZ5nVr4RqCM4Z5SGtv6QctCA9

**Comment by Szegoo** [comment]
Polkadot address: 126X27SbhrV19mBFawys3ovkyBS87SGfYwtwa8J2FjHrtbmA

**Comment by Szegoo** [comment]
> /tip small

It doesn't seem like this worked, only the first tip was submitted if I am not mistaken.

**Comment by sudipghimire533** [comment]
On my address as well there we no balance whatsoever. Maybe because mine was freshly created address and it didn't even had existential deposit

**Comment by Szegoo** [comment]
> On my address as well there we no balance whatsoever. Maybe because mine was freshly created address and it didn't even had existential deposit

Yours was submitted successfully, you can check it on https://www.dotreasury.com/dot/tips

**Comment by sudipghimire533** [comment]
https://www.dotreasury.com/dot/tips/11627463_0xeed4b3487ce0c15cb0cc66f0293eece4f70001ca6f75951d9b7880d8bce3822a


You can see: `value: 0` 
image

**Comment by Szegoo** [comment]
That is because no one has yet declared a tip value, you can check how this works in the `tips` pallet(https://github.com/paritytech/substrate/tree/master/frame/tips/)

**Comment by sudipghimire533** [comment]
> That is because no one has yet declared a tip value, you can check how this works in the `tips` pallet(https://github.com/paritytech/substrate/tree/master/frame/tips/)

Is that something I need to do or parity team?

**Comment by Szegoo** [comment]
> > That is because no one has yet declared a tip value, you can check how this works in the `tips` pallet(https://github.com/paritytech/substrate/tree/master/frame/tips/)
> 
> Is that something I need to do or parity team?

You don't have to do anything, you need to wait for the tippers to declare the tip value.

**Comment by athei** [comment]
This broke cumulus. How was that merged without a cumulus companion? Companion build broken?

**Pull request merged**